zcash_client_backend/src/keys.rs

```diff
@@ -74,12 +86,56 @@ pub enum DerivationError {
     Transparent(hdwallet::error::Error),
 }
 
+/// A version identifier for the encoding of unified spending keys.
+///
+/// Each epoch corresponds to a range of block heights during which the unified spending key parsed
+/// from an encoded form tagged with that epoch's identifier is expected to provide sufficient
+/// spending authority to spend any non-Sprout shielded note created in a transaction within the
+/// epoch's block range.
+#[derive(Debug, PartialEq, Eq)]
+pub enum Epoch {
+    /// The Orchard epoch begins at Orchard activation, and will end if a new pool that
+    /// requires a change to unified spending keys is introduced.
+    Orchard,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum DecodingError {
+    ReadError(&'static str),
+    EpochInvalid,
+    EpochMismatch(Epoch),
+    TypecodeInvalid,
+    LengthInvalid,
+    LengthMismatch(Typecode, u32),
+    InsufficientData(Typecode),
+    KeyDataInvalid(Typecode),
+}
+
+impl Epoch {
+    /// Returns the unique identifier for the epoch.
+    #[cfg(feature = "unstable")]
+    fn id(&self) -> u32 {
+        // We use the consensus branch id of the network upgrade that introduced a
+        // new USK format as the identifier for the epoch.
+        match self {
+            Epoch::Orchard => u32::from(BranchId::Nu5),
+        }
+    }
+
+    #[cfg(feature = "unstable")]
+    fn from_id(id: u32) -> Option<Self> {
+        BranchId::try_from(id).ok().and_then(|b| match b {
+            BranchId::Nu5 => Some(Epoch::Orchard),
+            _ => None,
+        })
+    }
+}
+
 /// A set of viewing keys that are all associated with a single
 /// ZIP-0032 account identifier.
 #[derive(Clone, Debug)]
 #[doc(hidden)]
 pub struct UnifiedSpendingKey {
-    account: AccountId,
```

Comment:
Should we put the account id in the encoding rather than omitting it from this type? How was it used?

Reply:
The inclusion of the account ID was a misfeature, since we would need at least an account ID and a seed fingerprint to correctly identify the relationship of a USK to a particular seed. The association with the account will be checked when the USK is used for spending, using information from the wallet database.